dhis2/dhis2-core
# feat: GeoJson import [DHIS2-7171]

### Summary 
Adds GeoJSON bulk and single import API.

Bulk import expects a `.geojson` file in the `POST`. Both `application/json` and `application/geo+json` are allowed.

Bulk update URL: `/api/organisationUnits/geometry`
Single update URL: `/api/organisationUnits/{uid}/geometry`

The bulk import can be used with zip or gzip files as well.

### Parameters
Both can be used with attribute target using the `attributeId` parameter. When omitted the geometry is stored in the `geometry` field of the `OrganisationUnit`.

Bulk can be used synchronous (default) or asynchronous (parameter `async=true`).

A bulk file can use different identifiers to link a GeoJSON feature with its geometry to a organization unit.
By default the feature `id` field is expected to have the OU UID.
If another property in the feature refers to the organization unit this is given by parameter, for example `geoJsonId=false&geoJsonProperty=code` refers to a member named `code` within the feature's `properties` object.
If the identity values provided by the GeoJSON file are not organisation unit UIDs this can be changed to name or code using the `orgUnitProperty` parameter, for example `orgUnitProperty=code`. 

Both single and bulk update support a `dryRun`. When used the import is processed but the final update of the organisation units is not performed. This naturally cannot produce errors that would first occur when running the update but otherwise will behave identical. 

### Report
The response is a report with a import statistic and potential conflicts that occurred during the import.
Like with data value import the errors are indicated by type and index. If same type of error occurs multiple times this results in a single conflict referring to all items that had the same problem by their index in the features collection.
If a bulk import is done asynchronously the immediate response contains a reference to `/api/system/tasks/...` where the report eventually can be found at `/api/system/taskSummaries/...`

### Automatic Testing
Controller based tests were added to test most of the features (parameter variations) and error cases of the bulk import.
As the single import is just a different entry to the same code two tests just make sure the single import works for geometry field and attribute usage.

Since the reports might be stored in redis they must be able to serialize to and from JSON. 
This is tested by a new test.

### Manual Testing
There are quite a number of scenarios to test.

General steps are these:
1. prepare a file to import
2. run import: `POST` to `/api/organisationUnits/geometry` with a file
   or, run import: `POST` to `/api/organisationUnits/{uid}/geometry` with a just the `geometry` value of a feature from a file
3. check if the report is as expected (non async)
   or, (async) check `/api/system/taskSummaries/GEOJSON_IMPORT/{your-id}` (similar link just with `tasks` instead of `taskSummaries` is returned for asynchronous `POST` requests

Some examples for variants:
* async import using names `/api/organisationUnits/geometry?geoJsonId=false&geoJsonProperty=name&orgUnitProperty=name&async=true`
* import using UIDs but provided by the feature's properties `id` field: `/api/organisationUnits/geometry?geoJsonId=false&geoJsonProperty=id`
* import with organisation unit UIDs provided in feature's `id` field using an attribute:  `/api/organisationUnits/geometry?attributeId={attr}`
